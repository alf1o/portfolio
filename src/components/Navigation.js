import React from 'react';
import PropTypes from 'prop-types';
import Hidden from '@material-ui/core/Hidden';
import Drawer from '@material-ui/core/Drawer';
import SwipeableDrawer from '@material-ui/core/SwipeableDrawer';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Divider from '@material-ui/core/Divider';
import { NavLink } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';

const drawerWidth = 240;

const styles = theme => ({
  container: {
    zIndex: 1,
    overflow: 'hidden',
    position: 'relative',
    display: 'flex',
    width: drawerWidth,
    [theme.breakpoints.up('md')]: {
      width: '100%'
    }
  },
  toolbar: theme.mixins.toolbar,
  drawerPaper: {
    width: drawerWidth,
    [theme.breakpoints.up('md')]: {
      position: 'relative',
      width: '100vw'
    }
  },
  list: {
    [theme.breakpoints.up('md')]: {
      display: 'flex',
      justifyContent: 'space-around'
    }
  },
  links: {
    [theme.breakpoints.up('md')]: {
      marginLeft: 0,
      marginRight: 0,
      paddingLeft: 0,
      paddingRight: 0,
      textAlign: 'center'
    }
  }
});

const routes = [
  {
    name: 'Home',
    path: '/'
  },
  {
    name: 'Projects',
    path: '/projects/Arcade Game'
  },
  {
    name: 'Contacts',
    path: '/contacts'
  },
  {
    name: 'About',
    path: '/about'
  }
];

/**
  Render the side menu for navigation.
  Use a swipeable drawer for mobile and a permanent drawer otherwise.
*/
// TODO: styles for active link
function Navigation({ classes, onDrawerSwipe, isOpen }) {
  const drawer = (
    <List
      className={classes.list}
    >
      {routes.map(route => (
        <ListItem key={route.name}
          component={NavLink}
          to={route.path}
          className={classes.links}
        >
          <ListItemText primary={route.name} />
        </ListItem>
      ))}
    </List>
  );

  return (
    <div className={classes.container}>
      <Hidden
        mdUp={true}
      >
        <SwipeableDrawer
          variant="temporary"
          anchor="left"
          open={isOpen}
          onOpen={onDrawerSwipe}
          onClose={onDrawerSwipe}
          classes={{
            paper: classes.drawerPaper
          }}
          ModalProps={{
            keepMounted: true
          }}
        >
          <div className={classes.toolbar} />
          <Divider />
          {drawer}
        </SwipeableDrawer>
      </Hidden>
      <Hidden
        smDown={true}
        implementation="css"
      >
        <Drawer
          anchor="top"
          variant="permanent"
          open={true}
          classes={{
            paper: classes.drawerPaper
          }}
        >
          {drawer}
        </Drawer>
      </Hidden>
    </div>
  );
}

Navigation.propTypes = {
  classes: PropTypes.object.isRequired,
  onDrawerSwipe: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired
};

export { Navigation };
export default withStyles(styles)(Navigation);
