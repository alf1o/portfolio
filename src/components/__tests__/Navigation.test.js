import React from 'react';
import Navigation from '../Navigation';
import { createShallow } from '@material-ui/core/test-utils';
import ListItemText from '@material-ui/core/ListItemText';
import SwipeableDrawer from '@material-ui/core/SwipeableDrawer';
import Drawer from '@material-ui/core/Drawer';

const routes = [
  {
    name: 'Home',
    path: '/'
  },
  {
    name: 'Projects',
    path: '/projects/Arcade Game'
  },
  {
    name: 'Contacts',
    path: '/contacts'
  },
  {
    name: 'About',
    path: '/about'
  }
];

describe('`Navigation`', () => {
  let props = {};
  let mountedNav;
  const shallow = createShallow();

  function nav() {
    if (!mountedNav) mountedNav = shallow(<Navigation {...props} />);
    return mountedNav;
  }

  beforeEach(() => {
    props = {
      classes: {
        container: '',
        toolbar: '',
        drawerPaper: '',
        list: '',
        links: ''
      },
      onDrawerSwipe: jest.fn(),
      isOpen: false
    };
    mountedNav = undefined;
  });

  it('should render a `ListItem` for every route', () => {
    const len = routes.length;
    // component renders 2 drawers
    expect(nav().dive().find('.Navigation-links-5').length).toBe(len * 2);
  });

  it('should pass to `ListItem` a `to` prop which is `route.path`', () => {
    const lis = nav().dive().find('.Navigation-links-5');
    const len = routes.length;
    lis.forEach((li, i) => {
      expect(li.props().to).toBe(routes[i % len].path);
    });
  });

  it('should pass to `ListItemText` a `primary` prop which is `route.name`', () => {
    const lits = nav().dive().find(ListItemText);
    const len = routes.length;
    lits.forEach((lit, i) => {
      expect(lit.props().primary).toBe(routes[i % len].name);
    });
  });

  describe('`SwipeableDrawer`', () => {
    it('should receive an `open` prop which is `props.isOpen`', () => {
      const swd = nav().dive().find(SwipeableDrawer);
      expect(swd.props().open).toBe(nav().props().isOpen);
    });

    it('should receive an `onOpen` prop which is `props.onDrawerSwipe`', () => {
      const swd = nav().dive().find(SwipeableDrawer);
      expect(swd.props().onOpen).toEqual(nav().props().onDrawerSwipe);
    });

    it('should receive an `onClose` prop which is `props.onDrawerSwipe`', () => {
      const swd = nav().dive().find(SwipeableDrawer);
      expect(swd.props().onClose).toEqual(nav().props().onDrawerSwipe);
    });

    it('should render the `nav`', () => {
      const swd = nav().dive().find(SwipeableDrawer);
      expect(swd.find('[component="nav"]').length).toBe(1);
    });
  });

  describe('`Drawer`', () => {
    it('should render the `nav`', () => {
      const drw = nav().dive().find(Drawer);
      expect(drw.find({ component: 'nav' }).length).toBe(1);
    });
  });

  it('should match snapshot', () => {
    expect(nav()).toMatchSnapshot();
  });
});
