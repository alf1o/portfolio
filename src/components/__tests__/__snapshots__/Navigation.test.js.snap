// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Navigation\` should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Navigation)
    classes={
        Object {
            "container": "",
            "drawerPaper": "",
            "links": "",
            "list": "",
            "toolbar": "",
          }
    }
    isOpen={false}
    onDrawerSwipe={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "classes": Object {
        "container": "Navigation-container-1",
        "drawerPaper": "Navigation-drawerPaper-3",
        "links": "Navigation-links-5",
        "list": "Navigation-list-4",
        "toolbar": "Navigation-toolbar-2",
      },
      "isOpen": false,
      "onDrawerSwipe": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "classes": Object {
          "container": "Navigation-container-1",
          "drawerPaper": "Navigation-drawerPaper-3",
          "links": "Navigation-links-5",
          "list": "Navigation-list-4",
          "toolbar": "Navigation-toolbar-2",
        },
        "isOpen": false,
        "onDrawerSwipe": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
